### PR TITLE
chore: enhance airdrop offchain transfer

### PR DESCRIPTION
**Pull request by NguyenHuy1812**
**What does this PR do?**

-   [x] Enhance transfer offchain token with mochi-pay-api
-   [x] Move logic handle balances to mochi-pay-api, keep activityLogs in mochi-api

### CODE REVIEW

File: pkg/entities/offchain_tip_bot.go [hunk 1]
```diff
@@ -57,53 +82,51 @@ func (e *Entity) TransferToken(req request.OffchainTransferRequest) ([]response.
...
-		return []response.OffchainTipBotTransferToken{}, errors.New(insufficientBalanceLog.FailReason)
-	}
+	senderBalance, err := e.svc.MochiPay.GetBalance(senderProfile.ID, req.Token)
 	if err != nil {
+		e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&insufficientBalanceLog)
 		e.log.Fields(logger.Fields{"token": req.Token, "user": req.Sender}).Error(err, "[entity.TransferToken] repo.OffchainTipBotUserBalances.GetUserBalanceByTokenID() failed")
 		return nil, err
 	}
-	if senderBalance.Amount == 0 {
+	amount := new(big.Int)
```
Comment (trkhoi) [inline pkg/entities/offchain_tip_bot.go:99]: can check bigint utils

File: pkg/entities/offchain_tip_bot.go [hunk 2]
```diff
@@ -57,53 +82,51 @@ func (e *Entity) TransferToken(req request.OffchainTransferRequest) ([]response.
...
+		e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&insufficientBalanceLog)
+		return nil, fmt.Errorf("failed to convert string to big int")
+	}
+
+	if amount.Cmp(big.NewInt(0)) != 1 {
 		e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&insufficientBalanceLog)
 		return []response.OffchainTipBotTransferToken{}, errors.New(insufficientBalanceLog.FailReason)
 	}
 	// case transfer all
+	floatVal := new(big.Float).SetInt(amount)
```
Comment (trkhoi) [inline pkg/entities/offchain_tip_bot.go:111]: should create function in utils to convert

**Review by trkhoi — CHANGES_REQUESTED** [review]

**Review by trkhoi — APPROVED** [review]

File: pkg/entities/offchain_tip_bot.go [hunk 3]
```diff
@@ -38,6 +39,30 @@ func (e *Entity) TransferToken(req request.OffchainTransferRequest) ([]response.
...
 	}
+	log.Status = consts.OffchainTipBotTrasferStatusFail
+	log.FailReason = consts.OffchainTipBotFailReasonGetProfileFailed
+	// get senderProfile, recipientProfiles by discordID
+	mpReqs := request.MochiPayTransferRequest{}
+	senderProfile, err := e.svc.MochiProfile.GetByDiscordID(req.Sender)
+	if err != nil {
+		e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&log)
+		return []response.OffchainTipBotTransferToken{}, errors.New(log.FailReason)
+	}
```
Comment (vhbien000) [inline pkg/entities/offchain_tip_bot.go:50]: line break

File: pkg/entities/offchain_tip_bot.go [hunk 4]
```diff
@@ -38,6 +39,30 @@ func (e *Entity) TransferToken(req request.OffchainTransferRequest) ([]response.
...
+	mpReqs.From = &request.Wallet{
+		ProfileGlobalId: senderProfile.ID,
+	}
+
+	for _, v := range req.Recipients {
+		profile, err := e.svc.MochiProfile.GetByDiscordID(v)
+		if err != nil {
+			e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&log)
+			return []response.OffchainTipBotTransferToken{}, errors.New(log.FailReason)
+		}
```
Comment (vhbien000) [inline pkg/entities/offchain_tip_bot.go:60]: line break

**Review by vhbien000 — APPROVED** [review]

File: pkg/entities/offchain_tip_bot.go [hunk 5]
```diff
@@ -57,53 +82,53 @@ func (e *Entity) TransferToken(req request.OffchainTransferRequest) ([]response.
...
 	}
+
+	if amount.Cmp(big.NewInt(0)) != 1 {
+		log.FailReason = consts.OffchainTipBotFailReasonInvalidAmount
+		e.repo.OffchainTipBotActivityLogs.CreateActivityLog(&log)
+		return []response.OffchainTipBotTransferToken{}, errors.New(log.FailReason)
+	}
+
 	// case transfer all
+	intVal, _ := util.BigIntToFloat(amount, float64(senderBalance.Data[0].Token.Decimal))
```
[the remark this answers is not in the document]
Comment (NguyenHuy1812) [reply pkg/entities/offchain_tip_bot.go:118]: // Accuracy describes the rounding error produced by the most recent
// operation that generated a Float value, relative to the exact value.
type Accuracy int8

// Constants describing the Accuracy of a Float.
const (
	Below Accuracy = -1
	Exact Accuracy = 0
	Above Accuracy = +1
)
we can ignore that cause we don't care too specific in this bigint

**Pull request merged**